crates/astria-sequencer-relayer/src/relayer/celestia_client/mod.rs

```diff
@@ -0,0 +1,654 @@
+            &self.signing_keys,
+        );
+
+        let blob_tx = new_blob_tx(&signed_tx, blobs.iter());
+
+        let tx_hash = self.broadcast_tx(blob_tx).await?;
+
+        loop {
+            tokio::time::sleep(Duration::from_secs(1)).await;
+            if let Some(height) = self.get_tx(tx_hash.clone()).await? {
+                // allow: the height is never negative.
+                #[allow(clippy::cast_sign_loss)]
+                return Ok(height as u64);
+            }
+        }
+    }
+
+    /// Makes queries to the Celestia app to retrieve pricing information, which is cached in
+    /// `state` on success.
+    ///
+    /// This should ideally only be called once during the session (at startup by
+    /// `CelestiaClientBuilder::try_build`), but will also be called if an attempt to submit blobs
+    /// fails.
+    async fn fetch_and_cache_cost_params(&mut self, state: Arc<super::State>) -> Result<(), Error> {
+        let gas_per_blob_byte = self.fetch_blob_params().await?.gas_per_blob_byte;
+        let tx_size_cost_per_byte = self.fetch_auth_params().await?.tx_size_cost_per_byte;
+        let min_gas_price = self.fetch_min_gas_price().await?;
+        let cost_params =
+            CelestiaCostParams::new(gas_per_blob_byte, tx_size_cost_per_byte, min_gas_price);
+        state.set_celestia_cost_params(cost_params);
+        Ok(())
+    }
+
+    async fn fetch_account(&mut self) -> Result<BaseAccount, Error> {
+        let request = QueryAccountRequest {
+            address: self.address.0.clone(),
+        };
+        let response = self.auth_query_client.account(request).await;
+        trace!(?response);
+        account_from_response(response)
+    }
+
+    async fn fetch_blob_params(&mut self) -> Result<BlobParams, Error> {
+        let response = self
+            .blob_query_client
+            .params(QueryBlobParamsRequest {})
+            .await;
+        trace!(?response);
+        response
+            .map_err(|status| Error::FailedToGetBlobParams(GrpcResponseError::from(status)))?
+            .into_inner()
+            .params
+            .ok_or_else(|| Error::EmptyBlobParams)
+    }
+
+    async fn fetch_auth_params(&mut self) -> Result<AuthParams, Error> {
+        let response = self
+            .auth_query_client
+            .params(QueryAuthParamsRequest {})
+            .await;
+        trace!(?response);
+        response
+            .map_err(|status| Error::FailedToGetAuthParams(GrpcResponseError::from(status)))?
+            .into_inner()
+            .params
+            .ok_or_else(|| Error::EmptyAuthParams)
+    }
+
+    async fn fetch_min_gas_price(&mut self) -> Result<f64, Error> {
+        let mut min_gas_price_client = MinGasPriceClient::new(self.grpc_channel.clone());
+        let response = min_gas_price_client.config(MinGasPriceRequest {}).await;
+        trace!(?response);
+        min_gas_price_from_response(response)
+    }
+
+    /// Returns the tx hash on success.
```

Comment:
```suggestion
    /// Returns the tx hash if the tx is successfully placed into the node's mempool.
```

Reply:
Done in [c2a1e13](https://github.com/astriaorg/astria/pull/963/commits/c2a1e1393a7b594125d574e454d2f76bee6cfcac).